src/sage/doctest/util.py

```diff
@@ -103,6 +103,68 @@ class Timer:
         {}
         sage: TestSuite(Timer()).run()
     """
+
+    def _pid_cpu_seconds(self, pid):
+        r"""
+        Parse the ``/proc`` filesystem to get the cputime of the given
+        ``pid``.
+
+        This also includes the times for child processes **that have
+        been ``wait()``ed for and terminated**. Specifically, pexpect
+        processes DO NOT fall into that category.
+        """
+        with open(f"/proc/{pid}/stat", "r") as statfile:
+            stats = statfile.read().split()
+
+        # man 5 proc (linux)
+        cputicks = sum( float(s) for s in stats[13:17] )
+
+        hertz = sysconf("SC_CLK_TCK")
+        return (cputicks / hertz)
+
+    def _quick_cputime(self):
+        r"""
+        A fast and replacement for ``sage.misc.misc.cputime``.
+
+        This is a "reliable" replacement (on Linux/BSD) that takes
+        subprocesses (particularly pexpect interfaces) into
+        account. The ``cputime()`` function from the ``misc`` module
+        can be passed ``subprocesses=True``, but this has a few
+        faults; mainly that it relies on each pexpect interface to
+        implement its own ``cputime()`` function. And most of our
+        pexpect interfaces either don't implement one, or implement
+        one in a way that requires the subprocess (being pexpected) to
+        be in perfect working condition -- that will often not be the
+        case at the end of a doctest line.
+
+        OUTPUT:
+
+        A float measuring the cputime in seconds of the sage process
+        and all its subprocesses.
+        """
+        # Start by using os.times() to get the cputime for sage itself
+        # and any subprocesses that have been wait()ed for and that
+        # have terminated.
+        cputime = sum( times()[:4] )
+
+        # Now try to get the times for any pexpect interfaces, since
+        # they do not fall into the category above.
+        from sage.interfaces.quit import expect_objects
+        for s in expect_objects:
+            S = s()
+            if S and S.is_running():
+                try:
+                    cputime += self._pid_cpu_seconds(S.pid())
+                except (ArithmeticError, LookupError, OSError,
+                        TypeError, ValueError):
```

Comment:
Maybe it's better to catch errors in `_pid_cpu_seconds` and raise a well specified exception there?

Also: a. why are these methods? and b. why not fix `cputime(...,subprocesses=True)` instead?

Reply:
> Maybe it's better to catch errors in `_pid_cpu_seconds` and raise a well specified exception there?

I can't remember why I did it this way in the first place, but my best guess is that raising the exception in `_pid_cpu_seconds` would involve a little bit more code for no added benefit. Instead of,

```python
except (ArithmeticError, LookupError, OSError,
                        TypeError, ValueError):
```

we'd have both,

```python
except (ArithmeticError, LookupError, OSError,
                        TypeError, ValueError):
    raise OSError
...
except OSError:
    ...
```

> Also: a. why are these methods?

Just for namespacing. They're not meant to be used outside of the `Timer` class, and the leading underscore makes them pseudo-private by python convention.

> and b. why not fix `cputime(...,subprocesses=True)` instead?

The new method does something fundamentally different, and doesn't always work. For the purposes of `--warn-long`, it's fine if we mis-account for a few CPU seconds on macOS, but I wouldn't want to have to justify breaking `cputime(subprocesses=True)` on macOS in whatever situations it happens to work right now.

Reply:
> Maybe it's better to catch errors in `_pid_cpu_seconds` and raise a well specified exception there?

I thought about this for a while and I think it's a better design. It will take me all day to rebuild sage but I've pushed the commit so if I'm real lucky, the CI will pass. In the process I documented where all of the various exceptions can come from.